# LAB 02.03 - Pandas

In [3]:
import numpy as np
import pandas as pd

## Task 1: Extract data

we have a dataframe of items with a price and, **sometimes**, with an extra column (`margin`).

You will have to complete a function that will **filter** the dataframe selecting the rows:

- whose price is > 100
- **OR** whose margin is >10, if the column `margin` is present

your function must return **A LIST** with the item ids of the selected rows

For instance, with the following dataframe

             price  category
    itemid                  
    56556   108.15       2.0
    73065    83.54       2.0
    36619   114.42       0.0
    73414    82.94       0.0
    13410   115.13       2.0
    66153    91.50       1.0
    77380    85.82       0.0
    73249    95.95       0.0
    11654   100.22       2.0
    11972    77.16       1.0

your must return the following list:

     [56556, 36619, 13410, 11654]
     
But if you get the following dataframe

             price  category  margin
    itemid                          
    39059    98.11       0.0   11.04
    19526    98.11       1.0   11.25
    78176    94.34       1.0   10.51
    50948   102.37       1.0   10.77
    12111    98.07       1.0    8.50
    56191    98.53       1.0   11.65
    38887    91.49       2.0   11.24
    77915   117.30       0.0    8.64
    55010    96.13       0.0    8.95
    45925    98.59       1.0   10.45
    
you must return the following list

    [39059, 19526, 78176, 50948, 56191, 38887, 77915, 45925]

In [5]:
def create_df(missing=False, n=10):
    itemid   = np.random.randint(100000, size=n)+1000
    category = np.random.randint(3, size=n)
    price    = np.round(np.random.normal(loc=100, scale=10, size=n),2)
    margin   = np.round(np.random.normal(loc=10, scale=1, size=n),2)
    
    if missing:
        nmissing = np.random.randint(len(price)//2)+2                                     
        price[np.random.permutation(len(price))[:nmissing]] = np.nan
    
    d = pd.DataFrame(np.r_[[price, category, margin]].T, index=itemid, columns=["price", "category", "margin"])
    d.index.name="itemid"
    if np.random.random()>.5:
        d = d[d.columns[:2]]
        
    return d

In [ ]:
d = create_df()
d

price  category
itemid                  
35181    92.38       1.0
36868   113.41       0.0
99109    84.07       1.0
20499    86.48       1.0
8527     92.03       1.0
68705    75.32       1.0
97462    86.73       2.0
22177   109.86       1.0
61973   102.49       2.0
90249   101.13       1.0

In [ ]:
def select_items(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    try:
      return df.loc[(df['margin']>10) | (df['price']>100)].index.to_list()
    except:
      return df[df['price']>100].index.to_list()

**manually check your answer**

In [ ]:
d = create_df()
d

price  category  margin
itemid                          
77538   103.26       2.0   10.78
67521    92.92       2.0   10.51
3998     94.95       1.0   10.26
59007   113.07       1.0    8.89
52774   102.74       1.0    9.23
80728    96.05       1.0    8.67
48348    90.14       2.0   10.34
89482    85.13       1.0   11.18
54701    98.58       1.0   10.25
80948   107.56       1.0   10.68

In [ ]:
d.loc[ (d['margin']>10) | (d['price']>100) ]

price  category  margin
itemid                          
77538   103.26       2.0   10.78
67521    92.92       2.0   10.51
3998     94.95       1.0   10.26
59007   113.07       1.0    8.89
52774   102.74       1.0    9.23
48348    90.14       2.0   10.34
89482    85.13       1.0   11.18
54701    98.58       1.0   10.25
80948   107.56       1.0   10.68

In [ ]:
select_items(d)

[77538, 67521, 3998, 59007, 52774, 48348, 89482, 54701, 80948]

## Task 2: Group statistics

Complete the following function so that it returns a dataframe with the average, max and min **prices** per category. 

For instance, for the following dataframe

             price  category  margin
    itemid                          
    17946    93.85       1.0   10.64
    61190    91.72       1.0    9.76
    39639   100.16       1.0   10.67
    17791   110.44       2.0    9.65
    7333    101.05       1.0    9.69
    77362   122.33       0.0   11.14
    92646   108.13       2.0   10.58
    27797    85.52       2.0   10.88
    31746    97.56       0.0    9.75
    12355   101.04       2.0    9.51
    
you should return the following dataframe

                 media  maximo  minimo
    categoria                         
    0         109.9450  122.33  97.56
    1          96.6950  101.05  91.72
    2         101.2825  110.44  85.52
    
observe that your result
- must not be a **multilevel** columnset.
- the column names and the index name must be **exactly** as in the example.
- the **index** must be of type **int**.

In [ ]:
def get_stats(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()

    # Identify categories
    values = []
    categories = []
    for category in df['category'].unique():
      d = df[df['category'] == category].loc[:, ['price']].describe()
      values.append(d.loc[['mean', 'max', 'min']].values.flatten())
      categories.append(int(category))
  
    result = pd.DataFrame(values, index=categories, columns=['media','maximo','minimo']).sort_index()
    result.index.name="categoria"
    return result

**manually check your answer**

In [ ]:
d = create_df()
d

price  category
itemid                  
2265    102.39       0.0
40433   110.55       0.0
80993    94.56       1.0
99996    81.42       0.0
53939    96.99       2.0
40812    90.62       1.0
94336    86.36       0.0
69850   111.51       0.0
68481   101.15       1.0
72404    99.36       1.0

In [ ]:
get_stats(d)

media  maximo  minimo
categoria                         
0          98.4460  111.51   81.42
1          96.4225  101.15   90.62
2          96.9900   96.99   96.99

## Task 3: Fill in missing data

Fill in the missing data in the **price** column with the following procedure:

- compute the mean and std of the avaialable prices
- sample from a normal distribution with the computed mean and std (see [`np.random.normal`])(https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html), as many samples as missing data
- substitute the missing values with the samples

For instance, for this input dataframe:

             price  category  margin
    itemid                          
    18922      NaN       1.0   10.32
    69500   121.25       1.0   10.22
    76442    90.25       1.0   12.60
    33863   106.51       0.0   10.26
    15904    95.87       1.0   11.51
    41946   103.47       2.0    9.85
    85451    93.08       2.0    9.56
    70028   116.68       1.0    9.11
    26860      NaN       2.0    9.71
    12807    91.48       0.0    9.77
    
your solution might be similar to this (not exactly the same as you will be sampling data):


                 price  category  margin
    itemid                              
    18922    97.441188       1.0   10.32
    69500   121.250000       1.0   10.22
    76442    90.250000       1.0   12.60
    33863   106.510000       0.0   10.26
    15904    95.870000       1.0   11.51
    41946   103.470000       2.0    9.85
    85451    93.080000       2.0    9.56
    70028   116.680000       1.0    9.11
    26860   103.294843       2.0    9.71
    12807    91.480000       0.0    9.77
    
    
**WARN**: your function must not modify the original dataframe, make a copy of the input dataframe, fill the values in the copy and return it.

**HINT**: use the [`isna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) method of a dataframe or a series.

In [46]:
d = create_df(missing=True)
d

price  category  margin
itemid                          
22864   109.93       0.0   11.48
29407      NaN       2.0   10.83
96290      NaN       2.0    9.37
44304   109.41       1.0   11.19
26428    91.18       1.0    7.93
72610      NaN       2.0   10.17
24973      NaN       1.0    9.28
98768    76.75       1.0   11.09
60549      NaN       2.0   10.81
28789      NaN       1.0    8.92

In [62]:
mu = d['price'].mean()
sigma = d['price'].std()
s = np.random.normal(mu, sigma, np.sum(pd.isnull(d['price'])))

d.loc[(d.isna()['price'] == True), 'price']  = s

In [63]:
np.sum(pd.isnull(d['price']))

0

In [65]:
d.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10 entries, 22864 to 28789
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   price     10 non-null     float64
 1   category  10 non-null     float64
 2   margin    10 non-null     float64
dtypes: float64(3)
memory usage: 320.0 bytes


In [75]:
def fillna(df):
    # make sure to make a copy in case you modify the original df
    df = df.copy()
    mu = df['price'].mean()
    sigma = df['price'].std()
    s = np.random.normal(mu, sigma, np.sum(pd.isnull(df['price'])))

    df.loc[(df.isna()['price'] == True), 'price']  = s
    return df

**check manually your code**

In [76]:
d = create_df(missing=True)
d

price  category
itemid                  
81387      NaN       2.0
93840      NaN       0.0
31655      NaN       1.0
19328   103.62       2.0
69576    94.28       2.0
82719    97.93       2.0
19986      NaN       0.0
43015   109.45       1.0
12139      NaN       2.0
96433    93.68       2.0

In [80]:
fillna(d)

price  category
itemid                      
81387   109.202388       2.0
93840    98.586910       0.0
31655   101.440403       1.0
19328   103.620000       2.0
69576    94.280000       2.0
82719    97.930000       2.0
19986    97.841031       0.0
43015   109.450000       1.0
12139    98.666926       2.0
96433    93.680000       2.0

In [79]:
d

price  category
itemid                  
81387      NaN       2.0
93840      NaN       0.0
31655      NaN       1.0
19328   103.62       2.0
69576    94.28       2.0
82719    97.93       2.0
19986      NaN       0.0
43015   109.45       1.0
12139      NaN       2.0
96433    93.68       2.0